### PR TITLE
 Add API and Reconciler for Tekton Result

### LINKED ISSUE #163: Feature Request:  Add "Tekton Results API" Addon
It would be great to be able to deploy [Tekton Results API](https://github.com/tektoncd/experimental/tree/master/results) via TektonAddon.

Once  tektoncd/experimental#652 is fixed, it should only need definitions in `/command/manager/kodata/resources/addons/` (and documentation).  I can create a pull request adding it, if desired.

### PR DESCRIPTION
**Pull request by sm43**
Closes #163 

Signed-off-by: Shivam Mukhade <smukhade@redhat.com>

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [ ] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/operator/blob/master/CONTRIBUTING.md) for more details._

# Release Notes

```
User can now install Tekton Results by creating a new CR TektonResult. The installation process is documented at  https://github.com/tektoncd/operator/blob/main/docs/tekton_results.md. This is enabled only for Kubernetes Platform.
```

### CODE REVIEW

File: config/openshift/role.yaml [hunk 1]
```diff
@@ -260,3 +260,21 @@ rules:
...
+  resources:
+    - subjectaccessreviews
+  verbs:
+    - create
+- apiGroups:
+    - results.tekton.dev
+  resources:
+    - '*'
+  verbs:
+    - '*'
```
Comment (savitaashture) [inline config/openshift/role.yaml:280]: I see same changes at `config/base/role.yaml` 
So if changes are at base folder then we don't require it in Openshift folder :thinking:

**Comment by nikhil-thomas** [comment]
Looks great 👍 

A few things needs to be added:
- release not in pr description
- a doc in `docs/` mentioning that TektonResults `will not be` auton installed like other components, as the TektonResults needs 2 secrets which end user should provide.

**Comment by nikhil-thomas** [comment]
@wlynch could you take a look. 🧑‍💻 

short recap:
- this pr adss a new clusterScoped CRD `TektonResults` 
- user will have to create an [instance of this CRD with name `result`](https://github.com/sm43/operator/blob/2998a5db64833138adea4aed6ea0a2ff195d9648/config/crs/kubernetes/result/operator_v1alpha1_result_cr.yaml) and operator will install latest available (present in operator repo) release.yaml of TektonReults
    - operator reconciler will only 1 instance (name: result) of the clusterwide CRD
- reconciler logic in short: [read TektonResults reelase.yaml](https://github.com/sm43/operator/tree/2998a5db64833138adea4aed6ea0a2ff195d9648/cmd/kubernetes/operator/kodata/tekton-result/0.2.0), change namespace to `spec.targetNamespace`, create resources on cluster
- at present, unlike other tekton components operator will not auto-install TektonResults, as TektonResults need 2 secrets to be created before install.

cc @sm43

**Comment by sm43** [comment]
/hold
to add unit tests for transformers

**Comment by sm43** [comment]
> Looks great +1
> 
> A few things needs to be added:
> 
>     * release not in pr description
> 
>     * a doc in `docs/` mentioning that TektonResults `will not be` auton installed like other components, as the TektonResults needs 2 secrets which end user should provide.

@nikhil-thomas addressed the comments.
also added a condition to check if Target Namespace in TektonResult CR is same as TektonPipeline CR, if different then installation would fail.

/hold cancel

**Comment by nikhil-thomas** [comment]
@sm43 let us keep TektonResults CRD and contrller available only for Kubernetes now.
For OpenShift we use payload (release.yaml) created in `github.com/openshift/tektoncd-<component-name>. eg: (https://github.com/openshift/tektoncd-pipeline/blob/release-v0.22.0/openshift/release/tektoncd-pipeline-v0.22.0.yaml)

As there is no midstream project for TektonResult, lets avoid adding it to openshift builds.

/hold make this patch kubernetes build specific

File: test/e2e/common/tektonresultdeployement_test.go [hunk 2]
```diff
@@ -0,0 +1,119 @@
+// +build e2e
```
Comment (wlynch) [inline test/e2e/common/tektonresultdeployement_test.go:1]: nit: tektonresultdeployment_test.go

File: docs/tekton_results.md [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+   -out cert.pem \
+   -days 365 \
+   -nodes \
+   -subj "/CN=tekton-results-api-service.${NAMESPACE}.svc.cluster.local" \
+   -addext "subjectAltName = DNS:tekton-results-api-service.${NAMESPACE}.svc.cluster.local"
+   # Create new TLS Secret from cert.
+   $ kubectl create secret tls -n ${NAMESPACE} tekton-results-tls \
+   --cert=cert.pem \
+   --key=key.pem
+   ```
```
Comment (wlynch) [inline docs/tekton_results.md:44]: I'm not super familiar with operator best practices, but would it be possible to automate these steps for the user if we find that they are not present at reconcile time?

For the TLS cert, are there existing examples for how we can hook into cert-manager or something similar?

File: test/resources/tektonresults.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+	return lastState, nil
+}
+
+// IsTektonResultReady will check the status conditions of the TektonResult and return true if the TektonResult is ready.
+func IsTektonResultReady(s *v1alpha1.TektonResult, err error) (bool, error) {
+	return s.Status.IsReady(), err
+}
+
+// AssertTektonResultCRReadyStatus verifies if the TektonResult reaches the READY status.
+func AssertTektonResultCRReadyStatus(t *testing.T, clients *utils.Clients, names utils.ResourceNames) {
```
Comment (wlynch) [inline test/resources/tektonresults.go:90]: it's a bit odd to me to have an exported func in a library that requires a `testing.T`. Can we unexport these and move them to the test file where they are used?

File: cmd/kubernetes/operator/kodata/tekton-result/0.2.0/release.yaml [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+      app: tekton-results-api
+  template:
+    metadata:
+      annotations:
+        cluster-autoscaler.kubernetes.io/safe-to-evict: "false"
+      labels:
+        app: tekton-results-api
+        app.kubernetes.io/name: tekton-results
+        app.kubernetes.io/component: api
+        results.tekton.dev/release: "v0.1.1"
```
[the remark this answers is not in the document]
Comment (sm43) [reply cmd/kubernetes/operator/kodata/tekton-result/0.2.0/release.yaml:247]: yep. i downloaded both but copied wrong one I guess 😅

File: test/resources/tektonresults.go (hunk 4 above)
Comment (sm43) [reply test/resources/tektonresults.go:90]: oh yeah make sense. It seems the for all components it is similar way. I will address this in a separate pr

File: docs/tekton_results.md (hunk 3 above)
Comment (sm43) [reply docs/tekton_results.md:44]: yeah that's the plan to automate this also, To start with we kept the creation of secrets a manual steps. so, user will create secrets and then TektonResult CR.
cc @nikhil-thomas

**Comment by sm43** [comment]
@wlynch the release.yaml for `v0.2.0` of Results seems to have `results.tekton.dev/release: devel` and version seems to be missing. Could you please check once? 
https://github.com/tektoncd/results/releases/tag/v0.2.0

File: docs/tekton_results.md (hunk 3 above)
Comment (nikhil-thomas) [reply docs/tekton_results.md:44]: > would it be possible to automate these steps for the user if we find that they are not present at reconcile time?
that is definitely the ideal ux the operator should deliver.
we will definitely handle in in upcoming iterations. 🧑‍💻 👍 

i have created an issue to track this: Automate creation prerequisites for TektonResult install #331

**Comment by wlynch** [comment]
> @wlynch the release.yaml for `v0.2.0` of Results seems to have `results.tekton.dev/release: devel` and version seems to be missing. Could you please check once?
> https://github.com/tektoncd/results/releases/tag/v0.2.0

https://github.com/tektoncd/results/issues/115 to track.

Beyond the recently opened issues, nothing else is jumping out at me, so LGTM.

/lgtm

**Pull request merged**